File: archaic/src/main/java/resonantinduction/archaic/fluid/tank/TankNetwork.java
package resonantinduction.archaic.fluid.tank;

import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.PriorityQueue;

import net.minecraft.tileentity.TileEntity;
import net.minecraftforge.fluids.FluidStack;
import resonantinduction.core.fluid.FluidDistributionetwork;
import resonantinduction.core.fluid.IFluidDistribution;
import calclavia.lib.utility.FluidUtility;

/**
 * Network that handles connected tanks
 * 
 * @author DarkGuardsman
 */
public class TankNetwork extends FluidDistributionetwork
{
	public TankNetwork()
	{
		super();
		animateDistribution = true;
	}

	@Override
	public void update()
	{
		// if (FMLCommonHandler.instance().getEffectiveSide() == Side.SERVER)
		// distributeConnectors();
	}

	@Override
	public void distributeConnectors()
	{
		int animateRate = 0;

		FluidStack totalFluid = getTank().getFluid();
		int lowestY = 255, highestY = 0;

		if (totalFluid == null || totalFluid.getFluid().isGaseous())
		{
			super.distributeConnectors();
		}
		else if (getConnectors().size() > 0)
		{
			FluidStack distributeFluid = totalFluid.copy();
			HashMap<Integer, Integer> heightCount = new HashMap<Integer, Integer>();
			PriorityQueue<IFluidDistribution> heightPriorityQueue = new PriorityQueue<IFluidDistribution>(1024, new Comparator()
			{
				@Override
				public int compare(Object a, Object b)
				{
					TileEntity wa = (TileEntity) a;
					TileEntity wb = (TileEntity) b;
					return wa.yCoord - wb.yCoord;
				}
			});

			for (IFluidDistribution connector : this.getConnectors())
			{
				if (connector instanceof TileEntity)
				{
					int yCoord = ((TileEntity) connector).yCoord;

					if (yCoord < lowestY)
					{
						lowestY = yCoord;
					}

					if (yCoord > highestY)
					{
						highestY = yCoord;
					}

					heightPriorityQueue.add(connector);
					heightCount.put(yCoord, heightCount.containsKey(yCoord) ? heightCount.get(yCoord) + 1 : 1);
				}
			}

			boolean didChange = false;
			//System.out.println("TANK UPDATE " + distributeFluid.amount);
			while (!heightPriorityQueue.isEmpty())
			{
				IFluidDistribution distributeNode = heightPriorityQueue.poll();
				int yCoord = ((TileEntity) distributeNode).yCoord;
				int connectorCount = heightCount.get(yCoord);

				if (distributeFluid == null || distributeFluid.amount <= 0)
				{
					break;
				}

				int fluidPer = (distributeFluid.amount / connectorCount) + (distributeFluid.amount % connectorCount);
				int deltaFluidAmount = (fluidPer - distributeNode.getInternalTank().getFluidAmount()) / 10;
				
				distributeNode.getInternalTank().setFluid(FluidUtility.getStack(distributeFluid,  fluidPer));
				
				/*
				 * 				System.out.println(connectorCount + " : " + fluidPer);

				if (deltaFluidAmount > 0)
				{
					distributeNode.getInternalTank().fill(FluidUtility.getStack(distributeFluid, deltaFluidAmount), true);
				}
				else
				{
					//TODO: This causes quite a lot of issues.
					FluidStack drained = distributeNode.getInternalTank().drain(Math.abs(deltaFluidAmount), true);
				}*/

				distributeFluid.amount -= distributeNode.getInternalTank().getFluidAmount();

				if (deltaFluidAmount != 0)
					didChange = true;

				if (connectorCount > 1)
					connectorCount--;

				heightCount.put(yCoord, connectorCount);
				distributeNode.onFluidChanged();
			}
		}
	}

	@Override
	public TankNetwork newInstance()
	{
		return new TankNetwork();
	}
}

File: src/main/scala/resonantinduction/core/grid/fluid/distribution/TankNetwork.scala
package resonantinduction.core.grid.fluid.distribution

import java.util.HashMap

import net.minecraft.tileentity.TileEntity
import net.minecraftforge.fluids.FluidStack
import resonant.lib.utility.FluidUtility

import scala.collection.mutable

/** Network that handles connected tanks
  *
  * @author DarkGuardsman */
class TankNetwork extends FluidDistributionGrid
{
  needsUpdate = true

  override def update()
  {
    val networkTankFluid: FluidStack = tank.getFluid
    var lowestY: Int = 255
    var highestY: Int = 0

    if (getNodes.size > 0)
    {
      var totalFluid: Int = if (networkTankFluid != null) networkTankFluid.amount else 0
      val heightCount: HashMap[Integer, Integer] = new HashMap[_, _]
      val heightPriorityQueue = new mutable.PriorityQueue[IFluidDistributor]()(Ordering.by(
        (a: IFluidDistributor, b: IFluidDistributor) =>
        {
          if (networkTankFluid != null && networkTankFluid.getFluid.isGaseous) return 0
          val wa: TileEntity = a.asInstanceOf[TileEntity]
          val wb: TileEntity = b.asInstanceOf[TileEntity]

          if (wa.yCoord > wb.yCoord)
            a
          else
            b
        }
      ))

      import scala.collection.JavaConversions._
      for (connector <- this.getNodes)
      {
        if (connector.isInstanceOf[TileEntity])
        {
          val yCoord: Int = (connector.asInstanceOf[TileEntity]).yCoord
          if (yCoord < lowestY)
          {
            lowestY = yCoord
          }
          if (yCoord > highestY)
          {
            highestY = yCoord
          }
          heightPriorityQueue.add(connector)
          heightCount.put(yCoord, if (heightCount.containsKey(yCoord)) heightCount.get(yCoord) + 1 else 1)
        }
      }
      var didChange: Boolean = false
      while (!heightPriorityQueue.isEmpty)
      {
        val distributeNode: IFluidDistributor = heightPriorityQueue.poll
        val yCoord: Int = (distributeNode.asInstanceOf[TileEntity]).yCoord
        var connectorCount: Int = heightCount.get(yCoord)
        if (totalFluid <= 0)
        {
          distributeNode.getInternalTank.setFluid(null)
          distributeNode.onFluidChanged
          continue //todo: continue is not supported
        }
        val fluidPer: Int = totalFluid / connectorCount
        val deltaFluidAmount: Int = fluidPer - distributeNode.getInternalTank.getFluidAmount
        val current: Int = distributeNode.getInternalTank.getFluidAmount
        if (deltaFluidAmount > 0)
        {
          val filled: Int = distributeNode.getInternalTank.fill(FluidUtility.getStack(networkTankFluid, deltaFluidAmount), false)
          distributeNode.getInternalTank.fill(FluidUtility.getStack(networkTankFluid, deltaFluidAmount / 10), true)
          totalFluid -= current + filled
        }
        else
        {
          val drain: FluidStack = distributeNode.getInternalTank.drain(Math.abs(deltaFluidAmount), false)
          distributeNode.getInternalTank.drain(Math.abs(deltaFluidAmount / 10), true)
          if (drain != null) totalFluid -= current - drain.amount
        }
        if (deltaFluidAmount != 0) didChange = true
        if (connectorCount > 1) ({connectorCount -= 1; connectorCount + 1 })
        heightCount.put(yCoord, connectorCount)
        distributeNode.onFluidChanged
      }
      if (!didChange) needsUpdate = false
    }
  }
}